Feat/mui themes 1816

tltowles opened the pull request:
Here I have bootstrapped one approach to exposing theme-data to the Material UI theme provider https://material-ui.com/customization/themes/

Eight MUI theme objects are generated based on the 8 available HIG themes. Each MUI theme object has theme-data from HIG (metadata and resolvedRoles) embedded in it for easy access in Consumer context. The MuiThemeProvider component simply ignores this extra data when passed to createMuiTheme(theme) to generate a HIG based MUI. This extra data will be useful when building out the rest of the component examples in the HIG acceptance app.

That being said, I have only done a thorough pass on buttons for all of the themes. Icons within buttons should also work but are not demonstrated in these examples yet. Avatar and Typography have been started but require a little more work. Obviously, how the Material components are setup within the acceptance app will be different from that of the HIG React components so there will be differences between these two sets of components both visually and in code.

For my team, this fork, as it exists, thus far would be useful due to the fact that baseline color values, fonts, and buttons are in place.

We may want this package to live outside of the theme-data folder in it's own folder so that package versions are independent.

scmm17 commented on packages/theme-data/README.md:
````diff
@@ -47,8 +47,23 @@ console.log(lightGrayMediumDensityTheme);
 // }
 ```
 
+### Access theme data as Material UI Theme
````
I have not reviewed the code in detail yet, but there is one change I would like to see. I think the Material UI theming code should be in a separate package, not part of `packages/theme-data`. The theme  data project is for defining the basic theming data, and there should be separate projects for each of the particular implementations of the themes, such as material UI. I suggest that this code go in `packages/theme-material-ui`.

wmui51 commented:
A couple of things to note: 

- The acceptance app is that it's usually more of an internal tool for our designers to check on the React components, so we may not want to put the Material UI components in there as well.
- Can you rebase your commits so that the history won't show that `Theme-Data` was touched, we don't want it to accidentally get a version bump since it was part of a `feat` commit.

tltowles commented:
I have rebased this work to no longer introduce changes to theme-data. For the time being I have changed the acceptance app to just have two links to Avatar and Button Mui components. There will be a need to have Mui components demonstrated for development work somewhere.  If not the acceptance app then where should we place these component demos?

scmm17 commented:
This PR is looking good. I'm OK with having limited MUI testing in the acceptance app. Having Button and Avatar should be sufficient. @wmui51 Do you think we need something to clearly indicate that these are not HIG React components?

scmm17 commented:
@wmui51 What can be done about the failing "ci/circleci: visual-test" test? It looks like a Sauce credentials problem. I don't know anything about that part of the system.

wmui51 commented:
@scmm17  Personally I think in the acceptance app it might be better to explicitly say `material-ui` rather than `mui` as I hear more folks refer to it as `material` rather than `mui`, also since the acceptance app is kind of a testing/playground of sorts as long I am working in the repo with it it could be weird to label something `mui` since it is the same as my last name.

To get the tests running, I will have to create a branch and merge the fork into it and push it up, that way since I am a "contributor" the tests will run, and the contributor will still get credit for the commits.

@tltowles Could you pull down the latest `development` branch and rebase your branch against that as it is out of date. I will also pull down the branch and do a little bit of testing as well.

scmm17 commented:
I agree, we should not use the abbriation `mui` for the reasons @wmui51  stated. Call it `material-ui` instead.

wmui51 commented:
@tltowles should `@material-ui` be included into the acceptance app's `package.json`? I am currently getting the following error when I try to run it to test.

`Module not found: Can't resolve '@material-ui/core/Avatar' in '/Users/muiw/dev/hig/acceptance/src/pages/mui'`

tltowles commented:
@wmui51 Yes, I will add material-ui/core to the acceptance app package.json

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -1,6 +1,8 @@
 import AvatarPage from "./pages/AvatarPage";
+import MuiAvatarPage from "./pages/mui/AvatarPage";
```
The imports are listed alphabetically, I think we should either keep that format or put all the `material-ui` imports together.

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -1,6 +1,8 @@
+import MuiAvatarPage from "./pages/mui/AvatarPage";
 import BannerPage from "./pages/BannerPage";
 import ButtonsPage from "./pages/ButtonsPage";
+import MuiButtonsPage from "./pages/mui/ButtonsPage";
```
same as above.

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -20,7 +22,9 @@ import TypographyPage from "./pages/TypographyPage";
 
 export default [
   { path: "/avatar/", component: AvatarPage, label: "Avatar" },
+  { path: "/mui/avatar/", component: MuiAvatarPage, label: "Mui Avatar" },
```
Much like the comment above, I think all the `material-ui` routes info should stay together.

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -20,7 +22,9 @@ import TypographyPage from "./pages/TypographyPage";
   { path: "/avatar/", component: AvatarPage, label: "Avatar" },
+  { path: "/mui/avatar/", component: MuiAvatarPage, label: "Mui Avatar" },
   { path: "/button/", component: ButtonsPage, label: "Button" },
+  { path: "/mui/buttons/", component: MuiButtonsPage, label: "Mui Buttons" },
```
see above.

wmui51 commented on packages/theme-material-ui/README.md:
````diff
@@ -0,0 +1,14 @@
+```js
+import lightGrayMediumDensityTheme from '@hig/theme-material-ui/build/mui/lightGrayMediumDensityTheme';
+
+console.log(lightGrayMediumDensityTheme);
````
We should probably add more to the readme on how to use the themes for `material-ui`.

tltowles replied:
I think as we get more components we should definitely do this.

wmui51 commented on packages/theme-material-ui/package.json:
```diff
@@ -0,0 +1,60 @@
+    "release": "hig-scripts-release"
+  },
+  "peerDependencies": {
+    "@hig/theme-context": "^3.0.0",
```
i don't think you need `@hig/theme-context` as part of this package, as I didn't see it used. also `@hig/theme-data` should be set to the latest version.

wmui51 commented on packages/theme-material-ui/package.json:
```diff
@@ -0,0 +1,60 @@
+    "@hig/scripts": "^0.1.2",
+    "mkdirp": "^0.5.1",
+    "dedent": "^0.7.0",
+    "@hig/semantic-release-config": "^0.1.0",
```
a bit of a nitpick, but `@hig/semantic-release-config` should be moved up with the other `@hig` dev dependencies.

wmui51 commented on packages/theme-material-ui/scripts/buildThemes.js:
```diff
@@ -0,0 +1,272 @@
+/*
```
this commented chunk seems to be leftover copy and paste from `@hig/theme-data`.

wmui51 commented on packages/theme-material-ui/scripts/buildThemes.js:
```diff
@@ -0,0 +1,272 @@
+import fs from "fs";
+import mkdirp from "mkdirp";
+import dedent from "dedent";
+import colorSchemes from "../../theme-data/src/colorSchemes";
```
we should reference the `@hig/theme-data` package rather than using the relative paths from the repo. there can be instances where people may download this package only and the relative paths will not work.

wmui51 commented on packages/theme-material-ui/package.json:
```diff
@@ -0,0 +1,60 @@
+{
+  "name": "@hig/theme-material-ui",
+  "version": "1.0.0-alpha",
```
This will have to be changed to `1.0.0` for this to be released.

dylankenneally commented:
These changes are looking great from my perspective @tltowles, thank you for the clear amount of effort you've put into this.

Adding @osinjoku (Osi Njoku, Autodesk Melbourne) & @jazzhaiku (Billy Lisner, Autodesk Melbourne) to the FYI list, as the 3 of us will be consuming this, and contributing to it if possible / required.

tltowles commented:
Ok this is ready for another look.

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -21,7 +23,9 @@ import TypographyPage from "./pages/TypographyPage";
 
 export default [
   { path: "/avatar/", component: AvatarPage, label: "Avatar" },
+  { path: "/mui/avatar/", component: MuiAvatarPage, label: "Mui Avatar" },
```
I think this needs to be removed

wmui51 commented on acceptance/src/routes.js:
```diff
@@ -21,7 +23,9 @@ import TypographyPage from "./pages/TypographyPage";
   { path: "/avatar/", component: AvatarPage, label: "Avatar" },
+  { path: "/mui/avatar/", component: MuiAvatarPage, label: "Mui Avatar" },
   { path: "/button/", component: ButtonsPage, label: "Button" },
+  { path: "/mui/buttons/", component: MuiButtonsPage, label: "Mui Buttons" },
```
This should probably be removed as well.

wmui51 commented:
@scmm17 I was wondering what level of documentation would you like to see in the `Readme.md` as it is currently pretty empty.

osinjoku commented:
Material UI v4.0.0 was just released ([hours ago](https://github.com/mui-org/material-ui/releases/tag/v4.0.0))

There are some [small API tweaks](https://material-ui.com/guides/migration-v3/#theme) to the theme object (I can help with this).

Does it make sense to match versions of `[@hig/theme-material-ui](https://github.com/Autodesk/hig/pull/1835/files#diff-44b63dd0aa80658aa56bc0add6732d19R3)` with major `@material-ui` versions?

@scmm17 @tltowles

scmm17 commented:
That's an interesting question about the material UI version. This component should use the version of Material UI that is in use by the teams consuming material UI, which I assume is V3 as this point. @tltowles Do you have an opinion on this?

scmm17 commented:
For documentation in the Readme file for this component, it should outline how to use it, as the other components do. This doesn't need to be extensive, but a short code sample would be good. It should also document which version of Material UI it works with.

osinjoku commented:
> For documentation in the Readme file for this component, it should outline how to use it, as the other components do. This doesn't need to be extensive, but a short code sample would be good. It should also document which version of Material UI it works with.

I've just tested these themes with `@material-ui` v4.0.0 and it works just fine. No need to change anything with theme generation. Two of the three [changes](https://material-ui.com/guides/migration-v3/#theme) were not relevant to what @tltowles wrote and the third,

```js
typography: {
  useNextVariants: true,
},
```
has no effect in v4.

scmm17 commented:
@tltowles This is looking really good, and I approve the submission. Now @wmui51 needs to have a look, and if he approves, he'll let you know what he needs to do to merge the PR. It's tricky because of our test environment.

wmui51 commented:
@tltowles I pulled this down to test and things look good. I noticed some visual discrepancies in acceptance app for the material ui buttons and the react buttons.
- The focus hovered states have different halos
- The `web light` theme buttons are quite different

I am also wondering if there should be a `storybook` for this for folks to look at or some gemini tests that could pick up any visual changes that occur especially since we may get a few contributions on this so making sure visual changes are intended would be good. While the visual-test may not run on circle, you can message me via slack for the gemini/sauce lab credentials so you can run the tests on your local.

@scmm17 do you think we should have this in `storybook` and `gemini` tests for this as well before merging in?

tltowles commented:
@wmui51 I fixed the focus hover issues and removed the avatar demo. The avatar page was not demonstrating anything from our material/hig generated themes anyways. I feel like visual tests make sense but storybook is more for functionality and prop selection right? Since these components are part of a separate library I would think they should remain outside of our storybook. At the moment I have no more bandwidth to keep going on this PR but I will circle back around when I get a chance if there are additional requirements.

scmm17 commented:
I think it's OK to not support a Storybook for now, but we can revisit that later, when Tim or somebody else has more time. I'm in favor of approving this soon so Tim can move on, and other teams, who have been expressing interest, can use it too.
Gemini tests are useful too, and should be added later, but are not needed immediately.

wmui51 commented:
@tltowles the changes to the `Button` look good and the states look like they match the React counterpart. If you have time can you pull down the latest development branch and rebase or merge the latest to it, that way I can pull the branch run the visual tests and merge it directly from this PR. One small thing I noticed was the `acceptance` app appears to have a new `yarn.lock` update when I ran `yarn`, if you can check that in as well that would be good.

If you don't have time for that, I can do those updates next week. Thanks again for working on this.

tltowles commented:
@wmui51 This should be ready for merge after CI runs. Thanks for all of the feedback!

wmui51 commented:
Going to merge this in, thanks for doing this. For now, I will be setting the `package.json` to `alpha` and update the documentation in another commit to let folks know that this is not the final product and leave it open for folks to contribute the rest of the components..
